Source/WebCore/rendering/RenderBox.cpp

```diff
@@ -402,18 +412,13 @@ void RenderBox::updateGridPositionAfterStyleChange(const RenderStyle& style, con
     if (!parentGrid)
         return;
 
-    if (oldStyle->gridItemColumnStart() == style.gridItemColumnStart()
-        && oldStyle->gridItemColumnEnd() == style.gridItemColumnEnd()
-        && oldStyle->gridItemRowStart() == style.gridItemRowStart()
-        && oldStyle->gridItemRowEnd() == style.gridItemRowEnd()
-        && oldStyle->order() == style.order()
-        && oldStyle->hasOutOfFlowPosition() == style.hasOutOfFlowPosition())
-        return;
-
     // Positioned items don't participate on the layout of the grid,
     // so we don't need to mark the grid as dirty if they change positions.
-    if (oldStyle->hasOutOfFlowPosition() && style.hasOutOfFlowPosition())
+    if ((oldStyle->hasOutOfFlowPosition() && style.hasOutOfFlowPosition()) || gridStyleHasNotChanged(style, oldStyle)) {
+        if (needsLayout() && isRenderGrid())
+            clearNeedsLayout();
```

Comment:
1. needs layout bit should only be cleared when we actually finished running layout on the renderer
2. while out-of-flow boxes do not participate in grid layout, they do participate in an ancestor _block_ layout . Only when we run _that_ layout, should this bit be cleared.
3. this bug is not specific to grid (see test case above)

Reply:
1. when the _dialog_ becomes modal, it gets "anchored" to the iCB -i.e. the containing block of such box is always the RenderView.
2. Containing block type of changes are normally driven by style mutations (e.g. `position: static` -> `position: fixed`)
3. Such changes are handled in styleWillChange/didChange where we mark all affected ancestors dirty

Now `dialog.showModal()` normally goes through the same codepath triggered by this selector in the UA stylesheet
```
dialog:modal {
    position: fixed;

```
unless there's a user set value on position (note the missing `!important`) -which is what we have here.
In such cases we don't see any position type of style change and we fail to mark the ancestors dirty (which means that the subsequent layout does not reach the dialog box -> stays dirty)

There are 2 options I can think of
- force `position: fixed` on elements in top layer (should check the spec if it's doable at all)
- mark renderers dirty where we can actually detect this change e.g.

```
@ -4014,6 +4014,7 @@ void RenderLayer::establishesTopLayerWillChange()
 
     if (auto* parentLayer = parent())
         parentLayer->removeChild(*this);
+    renderer().setNeedsLayout();
 }
```

Reply:
Thanks, for the detailed explanation and solutions. I will look into this issue again with all these info you provided.

Reply:
The reduced test case helps me focus on real issue. Correct me if i am wrong, I think there is something else from the back trace i observed so far: 

At the beginning i was so confused why would #2 option work, as there will be a styleWillChange/didChange on `dialog.showModal()` anyway (it produces a `StyleDifference::Layout` diff which will  `setNeedsLayoutAndPrefWidthsRecalc();`). 

After compared the back trace between before and after applied #2, i found that the difference is that, this `setNeedsLayoutAndPrefWidthsRecalc();` is happened within `RenderElement::styleDidChange` which is happened after `styleWillChange` (where the original style bits will be cleared) and before `updateFromStyle();` (where style bits restored). this will cause an issue within `RenderObject::markContainingBlocksForLayout`, that `hasOutOfFlowPosition` check is always false. So `ancestor->setPosChildNeedsLayoutBit(true);` did not happen during. This is also the reason why #2 works and eventually the render layout got handled. 

For this specific case, now I feel this is more like a gap for OOF position when we set layout dirty bit during styleDidChange, and we expecting this render element will be handled during layoutBlock. I am wondering should we take its old style into consideration during `setNeedsLayoutAndPrefWidthsRecalc()->markContainingBlocksForLayout();` if it is done in `styleDidChange` to see this renderElement was `wasOutOfFlowPositioned`? Or am I wrong that is desired behaviour?

```
if (diff == StyleDifference::Layout || diff == StyleDifference::SimplifiedLayout) {
        RenderCounter::rendererStyleChanged(*this, oldStyle, m_style);

        // If the object already needs layout, then setNeedsLayout won't do
        // any work. But if the containing block has changed, then we may need
        // to mark the new containing blocks for layout. The change that can
        // directly affect the containing block of this object is a change to
        // the position style.
        if (needsLayout() && oldStyle && oldStyle->position() != m_style.position())
            markContainingBlocksForLayout();

        if (diff == StyleDifference::Layout)
->          setNeedsLayoutAndPrefWidthsRecalc();
        else
            setNeedsSimplifiedNormalFlowLayout();
```

Besides above, #2 has below code path that current code doesnt: 
```
void RenderObject::markContainingBlocksForLayout(ScheduleRelayout scheduleRelayout, RenderElement* newRoot)
{
...
if (hasOutOfFlowPosition) {
            bool willSkipRelativelyPositionedInlines = !ancestor->isRenderBlock() || ancestor->isAnonymousBlock();
            // Skip relatively positioned inlines and anonymous blocks to get to the enclosing RenderBlock.
            while (ancestor && (!ancestor->isRenderBlock() || ancestor->isAnonymousBlock()))
                ancestor = ancestor->container();
            if (!ancestor || ancestor->posChildNeedsLayout())
                return;
            if (willSkipRelativelyPositionedInlines)
                container = ancestor->container();
->          ancestor->setPosChildNeedsLayoutBit(true);
            simplifiedNormalFlowLayout = true;
```

Reply:
#2 works because we mark the containing block chain dirty _before_ the element gets added to the top layer. Soon after it is part of the top layer, any function that tries to climb the ancestor chain is going to get the "anchored to the ICB" answer because both `containingBlock()` and `container()` functions consult that bit ignoring the position property or any other bit on the renderer.

Reply:
Got it, so this is expected. you also mentioned below which I did not quite understand previously. 
"unless there's a user set value on position (note the missing !important) -which is what we have here.
In such cases we don't see any position type of style change and we fail to mark the ancestors dirty"

Btw if we are going to apply #2 should we limited to isModal as popover will also have this code path if I remember correctly.